File: script-gen-manager/src/model/sweep_data/sweep_timing_config.rs
use serde::{Deserialize, Serialize};

use crate::{
    instr_metadata::base_metadata::BaseMetadata, model::sweep_data::parameters::ParameterInt,
};

use super::{
    number_limit::{NumberLimit, SmuTimingLimit},
    parameters::{ParameterFloat, ParameterString},
};

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct SweepTimingConfig {
    pub measure_count: ParameterInt,
    pub smu_timing: SmuTiming,
    pub psu_timing: PsuTiming,
}

impl SweepTimingConfig {
    pub fn new() -> Self {
        SweepTimingConfig {
            measure_count: ParameterInt::new("measureCount", 1),
            smu_timing: SmuTiming::new(),
            psu_timing: PsuTiming::new(),
        }
    }

    pub fn evaluate(&mut self) {
        self.smu_timing.evaluate();
        self.psu_timing.evaluate();
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct SmuTiming {
    pub nplc: ParameterFloat,
    pub aperture: ParameterFloat,
    pub source_auto_delay: ParameterString,
    pub source_delay: ParameterFloat,
    pub measure_auto_delay: ParameterString,
    pub measure_delay: ParameterFloat,
    pub nplc_type: ParameterString,
}

impl SmuTiming {
    pub fn new() -> Self {
        let mut smu_timing = SmuTiming {
            nplc: ParameterFloat::new("nplc", 1.0, None),
            aperture: ParameterFloat::new(
                "aperture",
                1e-6,
                Some(BaseMetadata::UNIT_SECONDS.to_string()),
            ),
            source_auto_delay: ParameterString::new("sourceAutoDelay"),
            source_delay: ParameterFloat::new(
                "sourceDelay",
                0.0,
                Some(BaseMetadata::UNIT_SECONDS.to_string()),
            ),
            measure_auto_delay: ParameterString::new("measureAutoDelay"),
            measure_delay: ParameterFloat::new(
                "measureDelay",
                0.0,
                Some(BaseMetadata::UNIT_SECONDS.to_string()),
            ),
            nplc_type: ParameterString::new("nplcType"),
        };
        smu_timing.set_defaults();
        smu_timing
    }

    pub fn set_defaults(&mut self) {
        self.source_auto_delay.range = vec![
            BaseMetadata::OFF_VALUE.to_string(),
            BaseMetadata::ON_VALUE.to_string(),
        ];
        self.source_auto_delay.value = BaseMetadata::OFF_VALUE.to_string();

        self.measure_auto_delay.range = vec![
            BaseMetadata::OFF_VALUE.to_string(),
            BaseMetadata::ON_VALUE.to_string(),
        ];
        self.measure_auto_delay.value = BaseMetadata::ON_VALUE.to_string();

        self.nplc_type.range = vec![String::from("NPLC"), String::from("Aperture")];
        self.nplc_type.value = String::from("NPLC");
    }

    pub fn evaluate(&mut self) {
        let smu_limits = SmuTimingLimit::new();
        //TODO: verify if additional validation is needed
        self.evaluate_nplc(&smu_limits.nplc_limits);
        self.evaluate_aperture(&smu_limits.aperture_limits);
        self.evaluate_source_delay(&smu_limits.source_delay_limits);
        self.evaluate_measure_delay(&smu_limits.measure_delay_limits);
    }

    // fn compute_effective_delay(&self) -> f64 {
    //     let mut delay = 0.0;
    //     if self.measure_auto_delay.value == BaseMetadata::OFF_VALUE.to_string() {
    //         delay += self.measure_delay.value;
    //     }
    //     if self.source_auto_delay.value == BaseMetadata::OFF_VALUE.to_string() {
    //         delay += self.source_delay.value;
    //     }
    //     delay
    // }

    // fn compute_t(&self) -> f64 {
    //     let mut meas_time_per_count = 0.0;
    // 	let mut effective_total_time = 0.0;
    //     //TODO: If we are using these computations, we need to obtain this value from the UI
    //     let line_freq = 60.0;

    //     meas_time_per_count = BaseMetadata::MIN_BUFFER_TIME + self.nplc.value * self.aperture.value / line_freq;
    //     //TODO: Should measure_count be a separate timing parameter for both SMU and PSU?
    //     effective_total_time = self.compute_effective_delay() + 1 * self.measure_count.value * meas_time_per_count;

    //     effective_total_time
    // }

    fn evaluate_nplc(&mut self, limit: &NumberLimit) {
        self.nplc.value = limit.limit(self.nplc.value);
    }

    fn evaluate_aperture(&mut self, limit: &NumberLimit) {
        self.aperture.value = limit.limit(self.aperture.value);
    }

    fn evaluate_source_delay(&mut self, limit: &NumberLimit) {
        self.source_delay.value = limit.limit(self.source_delay.value);
    }

    fn evaluate_measure_delay(&mut self, limit: &NumberLimit) {
        self.measure_delay.value = limit.limit(self.measure_delay.value);
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct PsuTiming {
    rate: ParameterString,
    aperture_value: Vec<f64>,
}

impl PsuTiming {
    pub fn new() -> Self {
        let mut psu_timing = PsuTiming {
            rate: ParameterString::new("rate"),
            aperture_value: vec![],
        };
        psu_timing.set_defaults();
        psu_timing
    }

    pub fn set_defaults(&mut self) {
        self.rate.range = vec![
            BaseMetadata::RATE_NORMAL.to_string(),
            BaseMetadata::RATE_FAST.to_string(),
        ];
        self.rate.value = BaseMetadata::RATE_NORMAL.to_string();
        self.aperture_value = vec![0.066667, 0.016667]; // 66.667ms, 16.667ms
    }

    pub fn evaluate(&mut self) {
        //let psu_limits = PsuTimingLimit::new();
        //TODO: verify if additional validation is needed
    }
}

File: script-gen-ui/src/app/model/sweep_data/SweepTimingConfig.ts
import {
  IParameterFloat,
  IParameterInt,
  IParameterString,
  IPsuTiming,
  ISmuTiming,
  ISweepTimingConfig,
} from '../interface';

export class ParameterInt {
  id: string;
  value: number;

  constructor(data: IParameterInt) {
    this.id = data.id;
    this.value = data.value;
  }

  toJSON() {
    return {
      id: this.id,
      value: this.value,
    };
  }
}

export class ParameterFloat {
  id: string;
  value: number;
  unit: string;

  constructor(data: IParameterFloat) {
    this.id = data.id;
    this.value = data.value;
    this.unit = data.unit;
  }

  toJSON() {
    return {
      id: this.id,
      value: this.value,
      unit: this.unit,
    };
  }
}

export class ParameterString {
  id: string;
  value: string;
  range: string[];

  constructor(data: IParameterString) {
    this.id = data.id;
    this.value = data.value;
    this.range = data.range;
  }

  toJSON() {
    return {
      id: this.id,
      value: this.value,
      range: this.range,
    };
  }
}

export class SweepTimingConfig {
  measure_count: ParameterInt;
  smu_timing: SmuTiming;
  psu_timing: PsuTiming;

  constructor(data: ISweepTimingConfig) {
    this.measure_count = new ParameterInt(data.measure_count);
    this.smu_timing = new SmuTiming(data.smu_timing);
    this.psu_timing = new PsuTiming(data.psu_timing);
  }

  toJSON() {
    return {
      measure_count: this.measure_count.toJSON(),
      smu_timing: this.smu_timing.toJSON(),
      psu_timing: this.psu_timing.toJSON(),
    };
  }
}

export class SmuTiming {
  nplc: ParameterFloat;
  aperture: ParameterFloat;
  source_auto_delay: ParameterString;
  source_delay: ParameterFloat;
  measure_auto_delay: ParameterString;
  measure_delay: ParameterFloat;
  nplc_type: ParameterString;

  constructor(data: ISmuTiming) {
    this.nplc = new ParameterFloat(data.nplc);
    this.aperture = new ParameterFloat(data.aperture);
    this.source_auto_delay = new ParameterString(data.source_auto_delay);
    this.source_delay = new ParameterFloat(data.source_delay);
    this.measure_auto_delay = new ParameterString(data.measure_auto_delay);
    this.measure_delay = new ParameterFloat(data.measure_delay);
    this.nplc_type = new ParameterString(data.nplc_type);
  }

  toJSON() {
    return {
      nplc: this.nplc.toJSON(),
      aperture: this.aperture.toJSON(),
      source_auto_delay: this.source_auto_delay.toJSON(),
      source_delay: this.source_delay.toJSON(),
      measure_auto_delay: this.measure_auto_delay.toJSON(),
      measure_delay: this.measure_delay.toJSON(),
      nplc_type: this.nplc_type.toJSON(),
    };
  }
}

export class PsuTiming {
  rate: ParameterString;
  aperture_value: number[];

  constructor(data: IPsuTiming) {
    this.rate = new ParameterString(data.rate);
    this.aperture_value = data.aperture_value;
  }

  toJSON() {
    return {
      rate: this.rate.toJSON(),
      aperture_value: this.aperture_value,
    };
  }
}
